### PR TITLE
SCP-48 User process survey

### PR DESCRIPTION
**Pull request by wenxi0504**
Logged-in user is able to see/edit/delete only their own surveys

### CODE REVIEW

File: server/controllers/survey.ts [hunk 1]
```diff
@@ -61,12 +61,21 @@ export function displayQuestionPage(req:Request, res: Response, next: NextFuncti
...
+        throw Error("Unreachable: this route handler is called only when the user is logged in");
+    }
+    const userId = req.user._id;
+    //survey.find() by using  survey id and owner(user id) to make sure the user edits own survey
+    Survey.find({ owner:userId, _id:id},(err)=> {
+        if(err){
             return next(err);
         }
-        res.render("index", { title: "EditSurvey", page: "editsurvey", surveyItem: survey });
+        getSurveyById(id, (err,survey) => {
```
Comment (JameeKim) [inline server/controllers/survey.ts]: You are making a `find` call on the database twice to get the same object. We do not want this to minimize database calls. I would suggest using `getSurveyById` only (and get rid of the `Survey.find` call above) and then checking if the `userId` is the same as the `owner` field of the survey object.

```ts
getSurveyById(id, (err, survey) => {
    if (err) {
        return next(err);
    }

    if (!userId.equals(survey.owner)) {
        return res.redirect("/account"); // redirect to the account page
        // OR you could also try sending out 401 Unauthorized error (more advanced)
    }

    // Now you can render the page here
});
```

File: server/controllers/survey.ts [hunk 2]
```diff
@@ -106,15 +115,25 @@ export function processMakeSurveyPage(req:Request, res: Response, next: NextFunc
...
-    Survey.findByIdAndRemove(id, {}, (err) => {
-        if(err)
-        {
+    // find the survey by survey id and user id
+    Survey.find({ owner:userId, _id:id},(err)=> {
+        if(err){
             return next(err);
         }
-        res.redirect("/account");
+        Survey.findByIdAndRemove(id, {}, (err) => {
```
Comment (JameeKim) [inline server/controllers/survey.ts]: Same as above, you're requesting 2 database operations. What you want to use is this: [`Survey.findOneAndRemove`](https://mongoosejs.com/docs/api/model.html#model_Model.findOneAndRemove).

```ts
Survey.findOneAndRemove({ owner: userId, _id: id }, (err) => {
    if (err) {
        return next(err);
    }

    res.redirect("/account");
    // This is done regardless of if the survey was actually found with the given id and owner id.
    // In business applications, you might want to do separate things depending on that.
    // In our case, I think it's fine to leave it like this.
});
```

**Review by JameeKim — CHANGES_REQUESTED** [review]
Thanks for the work! Please check the comments below.

In addition to individual comments below, I also need to mention that you need to add the check in `processEditSurveyPage` function, too. Try using [`Survey.findOneAndUpdate`](https://mongoosejs.com/docs/api/model.html#model_Model.findOneAndUpdate).

```ts
Survey.findOneAndUpdate({ owner: userId, _id: id }, updatedSurvey, {}, (err) => {
    if (err) {
        return next(err);
    }
    res.redirect("/account");
    // Same as one of the comments below, this redirects the user regardless of
    // whether the survey with the given id and owner id was found or not.
    // Not a big deal in our case, but ideally, we would want to return an error
    // if no such survey was found.
});
```

**Comment by wenxi0504** [comment]
Message received..I will fix it in the afternoon... Thanks

**Review by JameeKim — COMMENTED** [review]
Thanks for the work. One thing I want to mention is that `==` or `!=` in javascript is very... weird. You are better off using `===` or `!==` unless you exactly know what you're doing. I'll take care of it since it's a very tiny one. Again, thanks!

**Review by JameeKim — DISMISSED** [review]

**Review by JameeKim — APPROVED** [review]

**Pull request merged**